### PR TITLE
Add base for Concealed Perks

### PR DESCRIPTION
**Pull request by JoeFwd**
#284 
Non civilian items can be equipped when in civilian mode in the inventory screen.

When switching characters, or transferring items from both left and right inventories or resetting the transactions, the player should be able to equip the displayed character with the items described by the perk.

Multiple patches can be added to allow more items to be available in civilian mode if the displayed character has all perks.

If multiple patches describe the same items, there could be some collision issues. But this shouldn't occur anytime soon with the way perks are thought out.

This should be compatible with :
- **e1.3.1**.228624
- **e1.4.0**.229335
- Beta **e1.4.1**.229326

### CODE REVIEW

File: src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+using CommunityPatch;
+using HarmonyLib;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.CampaignSystem.ViewModelCollection;
+using TaleWorlds.Core;
+using TaleWorlds.Library;
+using static System.Reflection.BindingFlags;
+using static CommunityPatch.PatchApplicabilityHelper;
+
+namespace Patches {
```
Comment (Tyler-IN) [inline src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs:13]:
```suggestion
namespace CommunityPatch.Patches {
```

File: src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+      if (!IsTargetPatchable(AfterTransferMethodInfo, AfterTransferHashes)) {
+        return false;
+      }
+
+      return base.IsApplicable(game);
+    }
+
+    private static void MapLeftAndRightInventory(SPInventoryVM spInventoryVm, Action<SPItemVM> itemMapper) {
+      var rightItemListVm = (MBBindingList<SPItemVM>) RightItemListVmFieldInfo.GetValue(spInventoryVm);
+      var leftItemListVm = (MBBindingList<SPItemVM>) LeftItemListVmFieldInfo.GetValue(spInventoryVm);
```
Comment (Tyler-IN) [inline src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs:229]: Use `.BuildGetter<SPInventoryVM,MBBindingList<SPItemVM>>()`  for these two accesses, move to static readonly field.

File: src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+      leftItemListVm.ApplyActionOnAllItems(itemMapper);
+      rightItemListVm.ApplyActionOnAllItems(itemMapper);
+    }
+
+    private static void NonCivilianPerkItemsEquipableState(SPInventoryVM spInventoryVm, Func<SPItemVM, bool> isItemApplicableToPerk, bool state) {
+
+      Action<SPItemVM> makeCivilianItemUnequipable = item => {
+        var isCivilian = item?.ItemRosterElement.EquipmentElement.Item?.IsCivilian ?? false;
+        if (isItemApplicableToPerk(item) && !isCivilian) {
+          SetIsCivilianItemMethodInfo.Invoke(item, new object[] { state });
```
Comment (Tyler-IN) [inline src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs:244]: Is using `.BuildInvoker<Action<SPItemVM,bool>>()` and saving to static readonly possible here?

File: src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    
+    protected static void FlipEquipableStateOfNonCivilianPerkItemsIfDisplayedCharacterHasPerkOrNot(SPInventoryVM spInventoryVm, Func<SPItemVM, bool> isItemApplicableToPerk) {
+      var currentCharacter = (CharacterObject)CurrentCharacterFieldInfo.GetValue(spInventoryVm);
+      if (currentCharacter != null) {
+        FlipEquipableStateOfNonCivilianPerkItemsIfCharacterHasPerkOrNot(spInventoryVm, currentCharacter, isItemApplicableToPerk);
+      }
+    }
+    
+    protected static void StorePreviousCharacter(SPInventoryVM spInventoryVm, out CharacterObject currentCharacter) {
+      currentCharacter = (CharacterObject) CurrentCharacterFieldInfo.GetValue(spInventoryVm);
```
Comment (Tyler-IN) [inline src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs:278]: Try to use BuildGetter

File: src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+      if (CharacterHasPerk(previousCharacter)) {
+        MakeNonCivilianPerkItemsEquipable(spInventoryVm, isItemApplicableToPerk);
+      }
+      else {
+        MakeNonCivilianPerkItemsUnequipable(spInventoryVm, isItemApplicableToPerk);
+      }
+    }
+    
+    protected static void FlipEquipableStateOfNonCivilianPerkItemsIfDisplayedCharacterHasPerkOrNot(SPInventoryVM spInventoryVm, Func<SPItemVM, bool> isItemApplicableToPerk) {
+      var currentCharacter = (CharacterObject)CurrentCharacterFieldInfo.GetValue(spInventoryVm);
```
Comment (Tyler-IN) [inline src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs:271]: Try to use BuildGetter

File: src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        FlipEquipableStateOfNonCivilianPerkItemsIfCharacterHasPerkOrNot(spInventoryVm, currentCharacter, isItemApplicableToPerk);
+      }
+    }
+    
+    protected static void StorePreviousCharacter(SPInventoryVM spInventoryVm, out CharacterObject currentCharacter) {
+      currentCharacter = (CharacterObject) CurrentCharacterFieldInfo.GetValue(spInventoryVm);
+    }
+
+    protected static void KeepEquipableStateOfNonCivilianPerkItemsIfPreviousAndDisplayedCharacterHavePerkOrNot(SPInventoryVM spInventoryVm, CharacterObject previousCharacter, Func<SPItemVM, bool> isItemApplicableToPerk) {
+      var displayedCharacter = (CharacterObject) CurrentCharacterFieldInfo.GetValue(spInventoryVm);
```
Comment (Tyler-IN) [inline src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs:282]: Try to use BuildGetter

File: src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    public override IEnumerable<MethodBase> GetMethodsChecked() {
+      yield return UpdateCurrentCharacterIfPossibleMethodInfo;
+      yield return InitializeInventoryMethodInfo;
+      yield return AfterTransferMethodInfo;
+    }
+    
+    public override bool Applied { get; protected set; }
+
+    public override void Apply(Game game) {
+      CommunityPatchSubModule.Harmony.Patch(InitializeInventoryMethodInfo, postfix: new HarmonyMethod(InitializeInventoryPostfixMethodInfo));
```
Comment (Tyler-IN) [inline src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs:143]: Maybe `if (Applied) return;`

File: src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs (hunk 5 above)
Comment (JoeFwd) [reply src/CommunityPatch/Patches/EquipDraggedNonCivilianItemPatch.cs:271]: By trying out the following :

```cs
    internal static readonly FieldInfo CurrentCharacterFieldInfo = typeof(SPInventoryVM)
      .GetField("_currentCharacter", NonPublic | Instance | DeclaredOnly);
      
    internal static readonly Func<SPInventoryVM, CharacterObject> CurrentCharacterGetter = CurrentCharacterFieldInfo
      .BuildGetter<SPInventoryVM, CharacterObject>();
```

CurrentCharacterGetter retuns null during runtime. So going to use AccessTools.FieldRefAccess method instead.

**Pull request merged**